substrate/client/db/src/lib.rs

```diff
@@ -3112,6 +3141,62 @@ pub(crate) mod tests {
 		}
 	}
 
+	#[test]
+	fn displaced_leaves_after_finalizing_works_with_disconnect() {
+		// In this test we will create a situation that can typically happen after warp sync.
+		// The situation looks like this:
+		// g -> <unimported> -> a3 -> a4
+		// Basically there is a gap of unimported blocks at some point in the chain.
+
+		let backend = Backend::<Block>::new_test(1000, 100);
+		let blockchain = backend.blockchain();
+		let genesis_number = 0;
+		let genesis_hash =
+			insert_header(&backend, genesis_number, Default::default(), None, Default::default());
+
+		let a3_number = 3;
+		let a3_hash = insert_disconnected_header(
+			&backend,
+			a3_number,
+			H256::from([200; 32]),
+			H256::from([1; 32]),
+			true,
+		);
+
+		let a4_number = 4;
+		let a4_hash =
+			insert_disconnected_header(&backend, a4_number, a3_hash, H256::from([2; 32]), true);
+		{
+			let displaced =
+				blockchain.displaced_leaves_after_finalizing(a3_hash, a3_number).unwrap();
+			assert_eq!(displaced.displaced_leaves, vec![(genesis_number, genesis_hash)]);
+			assert_eq!(displaced.displaced_blocks, vec![]);
+		}
+
+		{
+			let displaced =
+				blockchain.displaced_leaves_after_finalizing(a4_hash, a4_number).unwrap();
+			assert_eq!(displaced.displaced_leaves, vec![(genesis_number, genesis_hash)]);
+			assert_eq!(displaced.displaced_blocks, vec![]);
+		}
+
+		// Import block a1 which has the genesis block as parent.
+		// g -> a1 -> <unimported> -> a3 -> a4
+		let a1_number = 1;
+		let a1_hash = insert_disconnected_header(
+			&backend,
+			a1_number,
+			genesis_hash,
+			H256::from([123; 32]),
+			false,
+		);
+		{
+			let displaced =
+				blockchain.displaced_leaves_after_finalizing(a3_hash, a3_number).unwrap();
+			assert_eq!(displaced.displaced_leaves, vec![(a1_number, a1_hash)]);
+			assert_eq!(displaced.displaced_blocks, vec![]);
+		}
```

Comment:
This assumes a specific narrow case of warp sync: we download future blocks and then sync the gap. In Subspace's Snap sync we actually have a similar situation, but *without* gap sync afterwards. We have to clear the gap manually due to Substrate's hardcoded notion of warp sync, see https://github.com/paritytech/polkadot-sdk/issues/4607 and issue linked from there. Also currently we are only able to do a single jump from genesis to a future block and no more jumps after that, which is another annoying limitation in Substrate that we'd like to lift if possible.

Another case that seems to be missing here is this:
```
// g -> a1 -> a2 -> <unimported> -> a3 -> a4
//  \-> b1 -> b2
```

If I understand proposed change correctly, it will remove both `a2` and `b2` leaves, while blocks themselves will remain in the database. Later when gap sync catches up, it'll connect to either `a2` or `b2` (or maybe some `c3` even, I am not familiar with warp sync logic), which means there will be blocks that will essentially "leak" and never be removed from the database due to them not being part of any of the active leafs and thus never discovered as displaced blocks by `displaced_leaves_after_finalizing` again.

The correct solution would be to simply delay the decision with two cases:
1. for warp sync it is sufficient to simply delay the decision until gap sync fills unimported gap and then decide what to prune with information of the full finalized chain
2. for other sync strategies like one used in Subspace we still need to delay the decision, but there will never be a decision, we can only wait for leaves that are 100% outside of pruning depth anyway, at which point the whole thing all the way down to genesis needs to be pruned unconditionally. I'm also wondering if it applies to warp sync if it is used on non-archival nodes (again, I'm not familiar with implementation, so let me know if warp sync is not applicable to non-archival nodes anyway).

For 2. above I think `displaced_leaves_after_finalizing` should return these leafs with "unknown" state such that they can be acted upon based on pruning logic. And pruning logic will need to be extended to support 2. of course.

Reply:
Defering the decision is fair. It is also what earlier implementations where doing. I changed the logic to not mark any leafs as displaced that have no current connection to the to-be-finalized block.

Regarding 2., I am not fully able to follow your explanation. I don't think it is in scope for this PR, which is only focusing on not making the node crash in case of finalization gap. This should restore feature parity to earlier implementations while maintaining the speed boosts from your improvements. We can then further iterate on it.

Reply:
Maybe not 100% in the scope, but definitely closely related. Simply imagine how would you prune blocks if gap is never synced after creation (because the chain doesn't need it).

Reply:
I am not saying we should not do it. I just want to have it in a follow up PR. I will not be able to work in the next days and would like to have this merged sooner rather than later due to the crashing. Once I am back I can take another look at it.